In [1]:
%matplotlib inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from bs4 import BeautifulSoup as bs
import re
from dateutil.parser import parse
from datetime import datetime
import pycountry_convert as pc


# Dataset 2 - Aviation Accident Database & Synopses, up to 2023 from NTSB - National Transportation Safety Board 

<div style="background-color: #9df9ef; padding: 10px;"></div>

## 2. Aviation Accident Database & Synopses, up to 2023

The dataset is [here](https://www.kaggle.com/datasets/khsamaha/aviation-accident-database-synopses)
The NTSB aviation accident database [here](https://www.ntsb.gov/Pages/home.aspx) contains information from 1962 and later about civil aviation accidents and selected incidents within the United States, its territories and possessions, and in international waters.

### 2.1. Reading, analyzing the dataset

In [2]:
avioset_ntsb = pd.read_csv('data/dataset_2_ntsb_gov/AviationData.csv', 
                           encoding='windows-1252', 
                           low_memory=False)

<div style="background-color: #9df9ef; padding: 10px;"></div>

### 2.2. Dataset cleaning, normalization

#### 2.2.1. Let's see what data do we have in all columns. Let's analyze additionally columns 6, 7, 28.

In [3]:
# 88889 rows × 31 columns
avioset_ntsb

Event.Id Investigation.Type Accident.Number  Event.Date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              Location        Country   Latitude   Longitude Airport.Code  \
0      MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1       BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2        Saltville, VA  United States  36.922223  -81.878056          NaN   
3           EUREKA, CA  United States        NaN         NaN          NaN   
4           Canton, OH  United States        NaN         NaN          NaN   
...                ...            ...        ...         ...          ...   
88884    Annapolis, MD  United States        NaN         NaN          NaN   
88885      Hampton, NH  United States        NaN         NaN          NaN   
88886       Payson, AZ  United States    341525N    1112021W          PAN   
88887       Morgan, UT  United States        NaN         NaN          NaN   
88888       Athens, GA  United States        NaN         NaN          NaN   

      Airport.Name  ... Purpose.of.flight         Air.carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      Total.Fatal.Injuries Total.Serious.Injuries Total.Minor.Injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      Total.Uninjured Weather.Condition  Broad.phase.of.flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
88884             0.0               NaN                    NaN   
88885             0.0    

In [4]:
avioset_ntsb.dtypes

Event.Id                   object
Investigation.Type         object
Accident.Number            object
Event.Date                 object
Location                   object
Country                    object
Latitude                   object
Longitude                  object
Airport.Code               object
Airport.Name               object
Injury.Severity            object
Aircraft.damage            object
Aircraft.Category          object
Registration.Number        object
Make                       object
Model                      object
Amateur.Built              object
Number.of.Engines         float64
Engine.Type                object
FAR.Description            object
Schedule                   object
Purpose.of.flight          object
Air.carrier                object
Total.Fatal.Injuries      float64
Total.Serious.Injuries    float64
Total.Minor.Injuries      float64
Total.Uninjured           float64
Weather.Condition          object
Broad.phase.of.flight      object
Report.Status 

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.2. Let's change column names like Event.Id into event_id.

In [5]:
# We'll make a new dataset so we don't lose the original
avioset_ntsb_low = avioset_ntsb

In [6]:
def change_col_names(df, old_symbol, new_symbol):
    df.columns = df.columns.str.lower().str.replace(old_symbol, new_symbol)
    return df

In [7]:
change_col_names(avioset_ntsb_low, '.', '_')

event_id investigation_type accident_number  event_date  \
0      20001218X45444           Accident      SEA87LA080  1948-10-24   
1      20001218X45447           Accident      LAX94LA336  1962-07-19   
2      20061025X01555           Accident      NYC07LA005  1974-08-30   
3      20001218X45448           Accident      LAX96LA321  1977-06-19   
4      20041105X01764           Accident      CHI79FA064  1979-08-02   
...               ...                ...             ...         ...   
88884  20221227106491           Accident      ERA23LA093  2022-12-26   
88885  20221227106494           Accident      ERA23LA095  2022-12-26   
88886  20221227106497           Accident      WPR23LA075  2022-12-26   
88887  20221227106498           Accident      WPR23LA076  2022-12-26   
88888  20221230106513           Accident      ERA23LA097  2022-12-29   

              location        country   latitude   longitude airport_code  \
0      MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1       BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2        Saltville, VA  United States  36.922223  -81.878056          NaN   
3           EUREKA, CA  United States        NaN         NaN          NaN   
4           Canton, OH  United States        NaN         NaN          NaN   
...                ...            ...        ...         ...          ...   
88884    Annapolis, MD  United States        NaN         NaN          NaN   
88885      Hampton, NH  United States        NaN         NaN          NaN   
88886       Payson, AZ  United States    341525N    1112021W          PAN   
88887       Morgan, UT  United States        NaN         NaN          NaN   
88888       Athens, GA  United States        NaN         NaN          NaN   

      airport_name  ... purpose_of_flight         air_carrier  \
0              NaN  ...          Personal                 NaN   
1              NaN  ...          Personal                 NaN   
2              NaN  ...          Personal                 NaN   
3              NaN  ...          Personal                 NaN   
4              NaN  ...          Personal                 NaN   
...            ...  ...               ...                 ...   
88884          NaN  ...          Personal                 NaN   
88885          NaN  ...               NaN                 NaN   
88886       PAYSON  ...          Personal                 NaN   
88887          NaN  ...          Personal  MC CESSNA 210N LLC   
88888          NaN  ...          Personal                 NaN   

      total_fatal_injuries total_serious_injuries total_minor_injuries  \
0                      2.0                    0.0                  0.0   
1                      4.0                    0.0                  0.0   
2                      3.0                    NaN                  NaN   
3                      2.0                    0.0                  0.0   
4                      1.0                    2.0                  NaN   
...                    ...                    ...                  ...   
88884                  0.0                    1.0                  0.0   
88885                  0.0                    0.0                  0.0   
88886                  0.0                    0.0                  0.0   
88887                  0.0                    0.0                  0.0   
88888                  0.0                    1.0                  0.0   

      total_uninjured weather_condition  broad_phase_of_flight  \
0                 0.0               UNK                 Cruise   
1                 0.0               UNK                Unknown   
2                 NaN               IMC                 Cruise   
3                 0.0               IMC                 Cruise   
4                 0.0               VMC               Approach   
...               ...               ...                    ...   
88884             0.0               NaN                    NaN   
88885             0.0    

In [8]:
avioset_ntsb_low.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

In [9]:
avioset_ntsb_low.columns

Index(['event_id', 'investigation_type', 'accident_number', 'event_date',
       'location', 'country', 'latitude', 'longitude', 'airport_code',
       'airport_name', 'injury_severity', 'aircraft_damage',
       'aircraft_category', 'registration_number', 'make', 'model',
       'amateur_built', 'number_of_engines', 'engine_type', 'far_description',
       'schedule', 'purpose_of_flight', 'air_carrier', 'total_fatal_injuries',
       'total_serious_injuries', 'total_minor_injuries', 'total_uninjured',
       'weather_condition', 'broad_phase_of_flight', 'report_status',
       'publication_date'],
      dtype='object')

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.3. Let's analyze nan in each column
- TODO

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.4. Let's look at the unique values in each column 
- TODO

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.5. Let's analyze columns 6, 7, 28 
- TODO

In [10]:
avioset_ntsb_low.columns[6]

'latitude'

In [11]:
avioset_ntsb_low.latitude.unique()

array([nan, '36.922223', '42.445277', ..., '321814N', '039101N',
       '373829N'], dtype=object)

In [12]:
avioset_ntsb_low.columns[7]

'longitude'

In [13]:
avioset_ntsb_low.latitude.unique()

array([nan, '36.922223', '42.445277', ..., '321814N', '039101N',
       '373829N'], dtype=object)

In [14]:
avioset_ntsb_low.columns[28]

'broad_phase_of_flight'

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.6. How many na we have? Should we replace or not? - TODO

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.7. Let's convert date columns into datetime - TODO

<div style="background-color: #9df9ef; padding: 10px;"></div>

#### 2.2.8. Let's take a look on the summary of statistics for numerical columns - It's good idea first to change some dtypes !!
- TODO
max: The maximum value in the data.
It is interesting to further analyze what data we have for Number.of.Engines == 0. There are aircraft without engines, commonly referred to as gliders or sailplanes. These aircraft are designed to fly without the need for an engine, relying instead on natural sources of lift, such as rising air currents (thermals), ridge lift, or wave lift, to stay airborne.

In [15]:
avioset_ntsb_low.describe()

number_of_engines  total_fatal_injuries  total_serious_injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       total_minor_injuries  total_uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

#### count: The number of non-null entries: 
We have a pretty good amount of non-null entries.
*****
#### mean: The average value of the data.
*****
#### std: The standard deviation, which measures the amount of variation or dispersion from the mean.
*****
#### min: The minimum value in the data.
It is interesting to further analyze what data we have for Number.of.Engines == 0.
There are aircraft without engines, commonly referred to as gliders or sailplanes. These aircraft are designed to fly without the need for an engine, relying instead on natural sources of lift, such as rising air currents (thermals), ridge lift, or wave lift, to stay airborne.
- TODO
*****
#### 25%: The 25th percentile (first quartile), which is the value below which 25% of the data fall.
*****
#### 50%: The 50th percentile (median), which is the middle value of the data.
*****
#### 75%: The 75th percentile (third quartile), which is the value below which 75% of the data fall.
*****
#### max: The maximum value in the data.
It's interesting to see further:
- How many injured people do we have? 
- How many accidents with injures do we have? 
- How are these values scattered Fatal.Injures 349 max, Total.Serious.Injures - 161, Total.Minor.Injures - 380?
- TODO

<div style="background-color: #9df9ef; padding: 10px;"></div>